### PR TITLE
feat: add request/response hooks to more http clients

### PR DESCRIPTION
**Pull request by nozik**
Same as the case in https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/62 - align the functionality of all http clients.

### CODE REVIEW

**Comment by nozik** [comment]
@ericmustin @arielvalentin @plantfansam As promised, the implementation is pretty much identical. I've followed the guidelines from the previous PR :)

**Comment by nozik** [comment]
@arielvalentin No worries :)

File: instrumentation/faraday/lib/opentelemetry/instrumentation/faraday/middlewares/tracer_middleware.rb [hunk 1]
```diff
@@ -41,6 +42,16 @@ def call(env)
 
           attr_reader :app
 
+          def config
+            Faraday::Instrumentation.instance.config
+          end
+
+          def safe_execute_hook(hook, *args)
```
[the remark this answers is not in the document]
Comment (plantfansam) [reply instrumentation/faraday/lib/opentelemetry/instrumentation/faraday/middlewares/tracer_middleware.rb:49]: @arielvalentin I'm not sure I fully understand 😬. Are you saying that an alternative to this implementation (`safe_execute_hook` inside of `TracerMiddleware` would be for the user to add a second Faraday middleware that extracts requests attributes, stashes them somewhere, and then this instrumentation would read off of that stash? Or did I miss the point entirely?

[the remark this answers is not in the document]
Comment (arielvalentin) [reply instrumentation/faraday/lib/opentelemetry/instrumentation/faraday/middlewares/tracer_middleware.rb:49]: Yes, that's exactly what I was exploring/meandering/musing about

[the remark this answers is not in the document]
Comment (arielvalentin) [reply instrumentation/faraday/lib/opentelemetry/instrumentation/faraday/middlewares/tracer_middleware.rb:49]: Thoroughly convinced. 🙇

File: instrumentation/restclient/lib/opentelemetry/instrumentation/restclient/patches/request.rb [hunk 2]
```diff
@@ -37,13 +47,21 @@ def trace_request # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
             OpenTelemetry::Trace.with_span(span) do
               OpenTelemetry.propagation.inject(processed_headers)
             end
+            safe_execute_hook(config[:request_hook], span, self) unless config[:request_hook].nil?
+
+            span
+          end
+
+          def trace_request # rubocop:disable Metrics/AbcSize
+            span = create_request_span
```
Comment (plantfansam) [inline instrumentation/restclient/lib/opentelemetry/instrumentation/restclient/patches/request.rb:56]: Any particular reason to break this out into its own method, besides tidiness? I'd generally prefer to keep it all in `trace_request`, but don't feel super strongly.

File: instrumentation/restclient/test/opentelemetry/instrumentation/restclient/instrumentation_test.rb [hunk 3]
```diff
@@ -111,4 +111,108 @@
...
+        _(span.attributes['http.method']).must_equal 'GET'
+        headers = span.attributes[headers_attribute]
+        _(headers).wont_be_nil
+        parsed_headers = JSON.parse(headers)
+        _(parsed_headers['User-Agent']).must_include 'rest-client'
+        _(span.attributes[response_body_attribute]).must_equal response_body
+      end
+    end
+
+    describe 'invalid hook - wrong number of args' do
```
Comment (plantfansam) [inline instrumentation/restclient/test/opentelemetry/instrumentation/restclient/instrumentation_test.rb:156]:
```suggestion
    describe 'when hook has wrong number of args' do
```

File: instrumentation/restclient/test/opentelemetry/instrumentation/restclient/instrumentation_test.rb [hunk 4]
```diff
@@ -111,4 +111,108 @@
...
+        ::RestClient.get('http://example.com/body')
+        _(exporter.finished_spans.size).must_equal 1
+        _(span.name).must_equal 'HTTP GET'
+        _(span.attributes['http.method']).must_equal 'GET'
+        error_messages = received_exceptions.map(&:message)
+        _(error_messages.all? { |em| em.start_with?('wrong number of arguments') }).must_equal true
+      end
+    end
+
+    describe 'invalid hooks - throws an error' do
```
Comment (plantfansam) [inline instrumentation/restclient/test/opentelemetry/instrumentation/restclient/instrumentation_test.rb:186]:
```suggestion
    describe 'when hooks throw errors' do
```

**Review by plantfansam — CHANGES_REQUESTED** [review]
This is looking great! Lots of inline comments but I think we're making progress!

File: instrumentation/restclient/lib/opentelemetry/instrumentation/restclient/patches/request.rb (hunk 2 above)
Comment (nozik) [reply instrumentation/restclient/lib/opentelemetry/instrumentation/restclient/patches/request.rb:56]: Yes, it gave me a linter error on a method too many lines

File: instrumentation/faraday/README.md [hunk 5]
```diff
@@ -30,6 +30,21 @@ OpenTelemetry::SDK.configure do |c|
 end
 ```
 
+### Configuration options
+
+```ruby
+OpenTelemetry::SDK.configure do |c|
+  c.use 'OpenTelemetry::Instrumentation::Faraday', {
+    request_hook: lambda { |span, request|
```
[the remark this answers is not in the document]
Comment (nozik) [reply instrumentation/faraday/README.md]: Done

File: instrumentation/faraday/README.md [hunk 6]
```diff
@@ -30,6 +30,21 @@ OpenTelemetry::SDK.configure do |c|
...
+      # Extract custom attributes from request
+    },
+    response_hook: lambda { |span, response|
```
[the remark this answers is not in the document]
Comment (nozik) [reply instrumentation/faraday/README.md:41]: In this case the upstream name is response, not env, correct me if I'm wrong

File: instrumentation/faraday/test/opentelemetry/instrumentation/faraday/middlewares/tracer_middleware_test.rb [hunk 7]
```diff
@@ -127,4 +129,103 @@
...
+        client.get('/body')
+        _(exporter.finished_spans.size).must_equal 1
+        _(span.name).must_equal 'HTTP GET'
+        _(span.attributes['http.method']).must_equal 'GET'
+        error_messages = received_exceptions.map(&:message)
+        _(error_messages.all? { |em| em.start_with?('wrong number of arguments') }).must_equal true
+      end
+    end
+
+    describe 'invalid hooks - throws an error' do
```
[the remark this answers is not in the document]
Comment (nozik) [reply instrumentation/faraday/test/opentelemetry/instrumentation/faraday/middlewares/tracer_middleware_test.rb]: done

File: instrumentation/faraday/test/opentelemetry/instrumentation/faraday/middlewares/tracer_middleware_test.rb [hunk 8]
```diff
@@ -127,4 +129,103 @@
...
+        _(span.attributes['http.method']).must_equal 'GET'
+        headers = span.attributes[headers_attribute]
+        _(headers).wont_be_nil
+        parsed_headers = JSON.parse(headers)
+        _(parsed_headers['traceparent']).wont_be_nil
+        _(span.attributes[response_body_attribute]).must_equal response_body
+      end
+    end
+
+    describe 'invalid hook - wrong number of args' do
```
[the remark this answers is not in the document]
Comment (nozik) [reply instrumentation/faraday/test/opentelemetry/instrumentation/faraday/middlewares/tracer_middleware_test.rb]: done

File: instrumentation/http/lib/opentelemetry/instrumentation/http/patches/client.rb [hunk 9]
```diff
@@ -25,14 +25,28 @@ def perform(req, options) # rubocop:disable Metrics/AbcSize
 
             tracer.in_span("HTTP #{request_method}", attributes: attributes, kind: :client) do |span|
               OpenTelemetry.propagation.inject(req.headers)
+              request_hook = instrumentation_config[:request_hook]
+              safe_execute_hook(request_hook, span, req) unless request_hook.nil?
```
[the remark this answers is not in the document]
Comment (nozik) [reply instrumentation/http/lib/opentelemetry/instrumentation/http/patches/client.rb]: done

File: instrumentation/http/lib/opentelemetry/instrumentation/http/patches/client.rb [hunk 10]
```diff
@@ -25,14 +25,28 @@ def perform(req, options) # rubocop:disable Metrics/AbcSize
...
               super.tap do |response|
                 annotate_span_with_response!(span, response)
+                response_hook = instrumentation_config[:response_hook]
```
[the remark this answers is not in the document]
Comment (nozik) [reply instrumentation/http/lib/opentelemetry/instrumentation/http/patches/client.rb]: done

File: instrumentation/http/test/instrumentation/http/patches/client_test.rb [hunk 11]
```diff
@@ -116,4 +116,108 @@
...
+        _(span.attributes['http.method']).must_equal 'GET'
+        headers = span.attributes[headers_attribute]
+        _(headers).wont_be_nil
+        parsed_headers = JSON.parse(headers)
+        _(parsed_headers['Traceparent']).wont_be_nil
+        _(span.attributes[response_body_attribute]).must_equal response_body
+      end
+    end
+
+    describe 'invalid hook - wrong number of args' do
```
[the remark this answers is not in the document]
Comment (nozik) [reply instrumentation/http/test/instrumentation/http/patches/client_test.rb]: done

File: instrumentation/http/test/instrumentation/http/patches/client_test.rb [hunk 12]
```diff
@@ -116,4 +116,108 @@
...
+        ::HTTP.get('http://example.com/body')
+        _(exporter.finished_spans.size).must_equal 1
+        _(span.name).must_equal 'HTTP GET'
+        _(span.attributes['http.method']).must_equal 'GET'
+        error_messages = received_exceptions.map(&:message)
+        _(error_messages.all? { |em| em.start_with?('wrong number of arguments') }).must_equal true
+      end
+    end
+
+    describe 'invalid hooks - throws an error' do
```
[the remark this answers is not in the document]
Comment (nozik) [reply instrumentation/http/test/instrumentation/http/patches/client_test.rb]: done

File: instrumentation/http_client/lib/opentelemetry/instrumentation/http_client/patches/client.rb [hunk 13]
```diff
@@ -12,6 +12,16 @@ module Patches
         module Client
           private
 
+          def config
```
[the remark this answers is not in the document]
Comment (nozik) [reply instrumentation/http_client/lib/opentelemetry/instrumentation/http_client/patches/client.rb]: done

File: instrumentation/http_client/test/instrumentation/http_client/patches/client_test.rb [hunk 14]
```diff
@@ -124,4 +124,114 @@
...
+        _(span.attributes['http.method']).must_equal 'GET'
+        headers = span.attributes[headers_attribute]
+        _(headers).wont_be_nil
+        parsed_headers = JSON.parse(headers)
+        _(parsed_headers['traceparent']).wont_be_nil
+        _(span.attributes[response_body_attribute]).must_equal response_body
+      end
+    end
+
+    describe 'invalid hook - wrong number of args' do
```
[the remark this answers is not in the document]
Comment (nozik) [reply instrumentation/http_client/test/instrumentation/http_client/patches/client_test.rb]: done

File: instrumentation/http_client/test/instrumentation/http_client/patches/client_test.rb [hunk 15]
```diff
@@ -124,4 +124,114 @@
...
+        http.get('http://example.com/body')
+        _(exporter.finished_spans.size).must_equal 1
+        _(span.name).must_equal 'HTTP GET'
+        _(span.attributes['http.method']).must_equal 'GET'
+        error_messages = received_exceptions.map(&:message)
+        _(error_messages.all? { |em| em.start_with?('wrong number of arguments') }).must_equal true
+      end
+    end
+
+    describe 'invalid hooks - throws an error' do
```
[the remark this answers is not in the document]
Comment (nozik) [reply instrumentation/http_client/test/instrumentation/http_client/patches/client_test.rb]: done

File: instrumentation/restclient/lib/opentelemetry/instrumentation/restclient/patches/request.rb [hunk 16]
```diff
@@ -18,7 +18,17 @@ def execute(&block)
 
           private
 
-          def trace_request # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+          def config
```
[the remark this answers is not in the document]
Comment (nozik) [reply instrumentation/restclient/lib/opentelemetry/instrumentation/restclient/patches/request.rb:21]: done

File: instrumentation/restclient/test/opentelemetry/instrumentation/restclient/instrumentation_test.rb (hunk 3 above)
Comment (nozik) [reply instrumentation/restclient/test/opentelemetry/instrumentation/restclient/instrumentation_test.rb]: done

File: instrumentation/restclient/test/opentelemetry/instrumentation/restclient/instrumentation_test.rb (hunk 4 above)
Comment (nozik) [reply instrumentation/restclient/test/opentelemetry/instrumentation/restclient/instrumentation_test.rb]: done

File: instrumentation/faraday/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (plantfansam) [reply instrumentation/faraday/README.md:41]: You're right; apologies!

File: instrumentation/restclient/lib/opentelemetry/instrumentation/restclient/patches/request.rb (hunk 2 above)
Comment (plantfansam) [reply instrumentation/restclient/lib/opentelemetry/instrumentation/restclient/patches/request.rb:56]: OK, sounds good. If you want to change back to being one large method and disable that rule, that's great; we can do in a follow-up PR though.

File: instrumentation/restclient/lib/opentelemetry/instrumentation/restclient/patches/request.rb [hunk 17]
```diff
@@ -18,14 +18,24 @@ def execute(&block)
...
+
+          def create_request_span # rubocop:disable Metrics/AbcSize
             http_method = method.upcase
             instrumentation_attrs = {
               'http.method' => http_method.to_s,
               'http.url' => OpenTelemetry::Common::Utilities.cleanse_url(url)
             }
-            config = RestClient::Instrumentation.instance.config
-            instrumentation_attrs['peer.service'] = config[:peer_service] if config[:peer_service]
+            instrumentation_config = RestClient::Instrumentation.instance.config
```
Comment (plantfansam) [inline instrumentation/restclient/lib/opentelemetry/instrumentation/restclient/patches/request.rb:37]: We should use `config` method here; we can do in follow-up PR

**Review by plantfansam — APPROVED** [review]

**Comment by arielvalentin** [comment]
@nozik we do not have the permissions to update the branch because your fork was made under the `helios` org. Please update your branch with the latest from upstream main and then we will merge this PR.

**Comment by nozik** [comment]
Thanks @arielvalentin @plantfansam ! Branch is now rebased

**Pull request merged**